### PR TITLE
RMC-233 Request IR via EQ

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
EQ will be sending us fulfilment requests via pubsub to enable requests for individual response.

# What has changed
* Add models, processor and tests for EQ fulfilments

# How to test?
Run with linked branch of AT's and docker dev

# Links
https://trello.com/c/GBhnJs8F/144-rmc-233-request-ir-via-eq-13
https://github.com/ONSdigital/census-rm-docker-dev/pull/60

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[RMC-233 Request IR via EQ (13)](https://trello.com/c/GBhnJs8F/144-rmc-233-request-ir-via-eq-13)

**Review by hbrace — APPROVED** [review]

File: main_test.go [hunk 1]
```diff
@@ -59,12 +59,56 @@ func TestMessageProcessing(t *testing.T) {
...
-		cfg.FulfilmentConfirmedTopic, cfg.FulfilmentConfirmedProject, cfg.FulfilmentRoutingKey))
+		cfg.FulfilmentConfirmedTopic, cfg.FulfilmentConfirmedProject, cfg.FulfilmentConfirmationRoutingKey))
 
 	t.Run("Test PPO fulfilment confirmation", testMessageProcessing(
 		`{"dateTime":"2019-08-03T14:30:01","caseRef":"12345678","productCode":"P_OR_H1","channel":"PPO","type":"FULFILMENT_CONFIRMED","transactionId":"92971ad5-c534-48af-a8b3-92484b14ceef"}`,
 		`{"event":{"type":"FULFILMENT_CONFIRMED","source":"RECEIPT_SERVICE","channel":"PPO","dateTime":"2019-08-03T14:30:01Z","transactionId":"92971ad5-c534-48af-a8b3-92484b14ceef"},"payload":{"fulfilmentInformation":{"caseRef":"12345678","fulfilmentCode":"P_OR_H1"}}}`,
-		cfg.FulfilmentConfirmedTopic, cfg.FulfilmentConfirmedProject, cfg.FulfilmentRoutingKey))
+		cfg.FulfilmentConfirmedTopic, cfg.FulfilmentConfirmedProject, cfg.FulfilmentConfirmationRoutingKey))
+	t.Run("Test EQ fulfilment request", testMessageProcessing(
```
Comment (NickGrantX) [inline main_test.go:69]: spacing 🎨

File: main_test.go [hunk 2]
```diff
@@ -59,12 +59,56 @@ func TestMessageProcessing(t *testing.T) {
...
+			"source" : "QUESTIONNAIRE_RUNNER",
+			"channel" : "EQ",
+			"dateTime" : "2011-08-12T20:17:46.384Z",
+			"transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93"
+		},
+		"payload" : {
+			"fulfilmentRequest" : {
+				"fulfilmentCode": "UACIT1",
+				"caseId" : "bbd55984-0dbf-4499-bfa7-0aa4228700e9",
+				"individualCaseId" : "8e8ebf71-d9c6-4efa-a693-ae24e7116e98",
```
Comment (NickGrantX) [inline main_test.go:82]: I'd kinda like to see a test without indiv case id?

File: main_test.go [hunk 3]
```diff
@@ -172,7 +216,7 @@ func TestStartProcessors(t *testing.T) {
 		return
 	}
 
-	assert.Len(t, processors, 5, "StartProcessors should return 5 processors")
+	assert.Len(t, processors, 6, "StartProcessors should return 5 processors")
```
Comment (NickGrantX) [inline main_test.go]: been updated to 6 now - log message should report that

File: models/eqFulfilment.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+type EqFulfilmentPayload struct {
+	FulfilmentRequest *FulfilmentRequest `json:"fulfilmentRequest" validate:"required"`
+}
+
+type FulfilmentRequest struct {
+	FulfilmentCode   string   `json:"fulfilmentCode" validate:"required"`
+	CaseID           string   `json:"caseId" validate:"required"`
+	IndividualCaseID string   `json:"individualCaseId"`
+	Contact          *Contact `json:"contact" validate:"required"`
```
Comment (NickGrantX) [inline models/eqFulfilment.go:18]: is it required or can it be null or not present?

File: models/eqFulfilment_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+				"caseId" : "bbd55984-0dbf-4499-bfa7-0aa4228700e9",
+				"individualCaseID": "c8e11403-9392-48e3-bfd5-5e8350ca1c4e",
+				"contact": {
+					"telNo":"+447890000000"
+				}
+			}
+		}
+	}`, true))
+
+	t.Run("Valid EQ fulfilment request without contact block", testEqFulfilmentValidate(
```
Comment (NickGrantX) [inline models/eqFulfilment_test.go]: can we have one without indiv ID too

File: processor/eqFulfilmentProcessor_test.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/stretchr/testify/assert"
+	"testing"
+	"time"
+)
+
+func TestConvertEqFulfilmentToRmMessage(t *testing.T) {
+	timeCreated, _ := time.Parse("2006-07-08T03:04:05Z", "2008-08-24T00:00:00Z")
+	eqFulfilment := models.EqFulfilment{
+		EqFulfilmentEvent: &models.RmEvent{
+			Type:          "FULFILMENT_CONFIRMED",
```
Comment (NickGrantX) [inline processor/eqFulfilmentProcessor_test.go]: I know this is only test data, but type should be `FULFILMENT_REQUESTED`

File: processor/eqFulfilmentProcessor_test.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+			Source:        "RECEIPT_SERVICE",
```
Comment (NickGrantX) [inline processor/eqFulfilmentProcessor_test.go]: I know this is only test data, but source should be `EQ_RUNNER`

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Minor comments

File: models/eqFulfilment.go (hunk 4 above)
Comment (AdamHawtin) [reply models/eqFulfilment.go:18]: RM actually fails to process it without at least an empty `contact` block

**Review by NickGrantX — APPROVED** [review]

**Pull request merged**